### PR TITLE
Cookie Tests

### PR DESCRIPTION
**Pull request by hughbe**
Added tests, especially for CookieContainer/Parser/Tokenizer/Collection

### CODE REVIEW

**Comment by hughbe** [comment]
Migrated from https://github.com/dotnet/corefx/pull/4851

**Comment by davidsh** [comment]
@stephentoub @CIPop PTAL

File: src/System.Net.Primitives/tests/FunctionalTests/CookieTest.cs [hunk 1]
```diff
@@ -337,10 +317,10 @@ public static void Equals_Compare_Success()
         }
 
         [Fact]
-        public static void ToString_Compare_Success()
+        public static void Tostring_Compare_Success()
```
Comment (justinvp) [inline src/System.Net.Primitives/tests/FunctionalTests/CookieTest.cs]: Tostring -> ToString

File: src/System.Net.Primitives/tests/FunctionalTests/CookieCollectionTest.cs [hunk 2]
```diff
@@ -121,6 +121,8 @@ public static void IndexSubscript_Get_Success()
             Assert.Equal(cc["name1"], c1);
             Assert.Equal(cc["name2"], c2);
             Assert.Equal(cc["name3"], c4);
+
+            Assert.Null(cc["no such name"]);
```
Comment (CIPop) [inline src/System.Net.Primitives/tests/FunctionalTests/CookieCollectionTest.cs]: I don't understand this change. The "Success" in the name means that you are able to extract the cookie. This line asserts that you cannot.
Please undo this change from this and below test.

Comment (hughbe) [inline src/System.Net.Primitives/tests/FunctionalTests/CookieCollectionTest.cs]: @CIPop  fixed

File: src/System.Net.Primitives/tests/FunctionalTests/CookieContainerTest.cs [hunk 3]
```diff
@@ -106,130 +182,440 @@ public static void PerDomainCapacity_Set_Invalid()
...
+            Cookie c1 = new Cookie("name1", "value", "", ".domain1.com");
+            Cookie c2 = new Cookie("name2", "value", "", ".domain2.com");
+            Cookie c3 = new Cookie("name3", "value", "", ".domain3.com");
+            c1.Expires = DateTime.Now.AddSeconds(1); // The cookie will expire in 1 second
+
+            CookieContainer cc = new CookieContainer(2);
+            cc.Add(c1);
+            cc.Add(c2);
+
+            Thread.Sleep(2000); // Sleep for 2 seconds to wait for the cookie to expire
```
Comment (CIPop) [inline src/System.Net.Primitives/tests/FunctionalTests/CookieContainerTest.cs]: Please use Task.Delay(2000) instead.

File: src/System.Net.Primitives/tests/FunctionalTests/CookieContainerTest.cs [hunk 4]
```diff
@@ -106,130 +182,440 @@ public static void PerDomainCapacity_Set_Invalid()
...
+            Assert.Equal(expected.Length, cc2.Count);
+
+            for (int i = 0; i < expected.Length; i++)
+            {
+                Cookie c1 = expected[i];
+                Cookie c2 = cc2[i];
+                Assert.Equal(c1.Name, c2.Name); // Primitive check for equality
+                Assert.Equal(c1.Value, c2.Value);
+            }
+
```
Comment (CIPop) [inline src/System.Net.Primitives/tests/FunctionalTests/CookieContainerTest.cs]: Nit: Please remove extra whitespace.

File: src/System.Net.Primitives/tests/FunctionalTests/CookieContainerTest.cs [hunk 5]
```diff
@@ -106,130 +182,440 @@ public static void PerDomainCapacity_Set_Invalid()
...
+            Cookie c2 = new Cookie("name2", "value", "", ".url2.com");
+            c1.Expires = DateTime.Now.AddSeconds(1); // The cookie will expire in 1 second
+
             CookieContainer cc = new CookieContainer();
+            cc.Add(c1);
+            cc.Add(c2);
 
-            Assert.Throws<ArgumentNullException>(() => cc.Add(null, new CookieCollection())); //Null uri
-            Assert.Throws<ArgumentNullException>(() => cc.Add(new Uri("http://contoso.com"), (CookieCollection)null)); //Null collection
+            Thread.Sleep(2000); // Let the thread sleep for 2 seconds, until the cookie is expired
```
Comment (CIPop) [inline src/System.Net.Primitives/tests/FunctionalTests/CookieContainerTest.cs]: Same as above: please switch to Task.Delay.

File: src/System.Net.Primitives/tests/FunctionalTests/CookieContainerTest.cs [hunk 6]
```diff
@@ -106,130 +182,440 @@ public static void PerDomainCapacity_Set_Invalid()
...
+        [Theory]
+        [MemberData("InvalidCookies")]
+        public static void Add_VerificationFailedCookie_Throws(Cookie c)
+        {
+            CookieContainer cc = new CookieContainer();
+            Assert.Throws<CookieException>(() => cc.Add(c));
         }
 
         [Fact]
         public static void Add_CookieCollection_Invalid()
```
Comment (CIPop) [inline src/System.Net.Primitives/tests/FunctionalTests/CookieContainerTest.cs]: The name could be enhanced: maybe `Add_CookieCollection_Null_Throws`?

File: src/System.Net.Primitives/tests/FunctionalTests/project.json [hunk 7]
```diff
@@ -2,7 +2,8 @@
   "dependencies": {
     "System.Diagnostics.Contracts": "4.0.0",
     "System.Runtime": "4.0.20",
-    "System.Threading.Tasks": "4.0.10",
+    "System.Threading.Tasks": "4.0.11-rc2-23610",
```
Comment (CIPop) [inline src/System.Net.Primitives/tests/FunctionalTests/project.json]: These need to remain generic versions (please remove the  `-rc2-23610` part) to ensure we're testing against the latest packages.
Also, any reason (newly required public API) you are changing them to 4.0.11?

**Comment by CIPop** [comment]
The project.json change needs to be addressed otherwise just nitpicks.

LGTM

Thanks for enhancing the tests @hughbe!

**Comment by hughbe** [comment]
@CIPop anytime, happy to help!

I've address your comments, cheers.
